src/nbl/ext/TextRendering/TextRendering.cpp

```diff
@@ -0,0 +1,250 @@
+
+using namespace nbl;
+using namespace nbl::core;
+using namespace nbl::asset;
+using namespace nbl::video;
+
+#include "nabla.h"
+#include <nbl/ext/TextRendering/TextRendering.h>
+
+namespace nbl
+{
+namespace ext
+{
+namespace TextRendering
+{
+
+uint32_t getCharacterAtlasPositionIx(char character)
+{
+	return int(character) - int(' ');
+}
+
+// Generates atlas of MSDF textures for each ASCII character
+FontAtlas::FontAtlas(IGPUQueue* queue, ILogicalDevice* device, const std::string& fontFilename, uint32_t msdfWidth, uint32_t msdfHeight, uint32_t charsPerRow)
+{
+	auto error = FT_Init_FreeType(&library);
+	assert(!error);
+
+	error = FT_New_Face(library, "C:\\Windows\\Fonts\\arialbd.ttf", 0, &face);
+	assert(!error);
+
+	auto fence = device->createFence(static_cast<nbl::video::IGPUFence::E_CREATE_FLAGS>(0));
+	auto commandPool = device->createCommandPool(queue->getFamilyIndex(), nbl::video::IGPUCommandPool::ECF_NONE);
+	nbl::core::smart_refctd_ptr<nbl::video::IGPUCommandBuffer> commandBuffer;
+	device->createCommandBuffers(commandPool.get(), nbl::video::IGPUCommandBuffer::EL_PRIMARY, 1u, &commandBuffer);
+
+	std::vector<nbl::core::smart_refctd_ptr<video::IGPUBuffer>> dataBuffers;
+	std::vector<asset::IImage::SBufferCopy> bufferCopies;
+
+	uint32_t characterCount = 0;
+	{
+		uint32_t atlasGlyphX = 0;
+		uint32_t atlasRow = 0;
+
+		for (char k = ' '; k <= '~'; k++)
+		{
+			wchar_t unicode = wchar_t(k);
+			uint32_t glyphIndex = FT_Get_Char_Index(face, unicode);
+
+			// special case for space as it seems to break msdfgen
```

Comment:
I'm no font expert but I think space is handled via `advanceX`, no glyph is needed. we can just simply advance the "pen position" and the next glyph would have a higher localXPosition or something.

Please do a little research to check if I didn't miss anything.

Reply:
it's not about layout, this if will skip characters that don't have any glyph associated with them, which would normally include space, but for some reason it looks like space is an empty/invalid glyph instead, which hits an assert in msdfgen when you feed it the space glyph

Reply:
So you're saying msdfgen should skip and ignore "space" instead of hitting an assert?

Maybe that's expected because space shouldn't have any glyph associated with it?